pkg/workflows/dontime/plugin.go

```diff
@@ -0,0 +1,256 @@
+			unscheduledFinishedExecutionIDs = append(unscheduledFinishedExecutionIDs, id)
+		}
+	}
+
+	// Collect up to batchSize unexpired requests
+	requests := map[string]int64{} // Maps executionID --> seqNum
+	for batchOffset := 0; batchOffset < p.store.requests.Len() && len(requests) < p.batchSize; {
+		batch, err := p.store.requests.RangeN(batchOffset, p.batchSize)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get request batch: %w", err)
+		}
+		if len(batch) == 0 {
+			break
+		}
+
+		timeoutCheck := time.Now()
+		for _, req := range batch {
+			if req.ExpiryTime().Before(timeoutCheck) {
+				// Request has been sitting in queue too long
+				p.store.requests.Evict(req.WorkflowExecutionID)
+				req.SendTimeout(ctx)
+				continue
+			}
+
+			// Validate request sequence number
+			numObservedDonTimes := 0
+			times, ok := previousOutcome.ObservedDonTimes[req.WorkflowExecutionID]
+			if ok {
+				// We have seen this workflow before so check against the sequence
+				numObservedDonTimes = len(times.Timestamps)
+			}
+
+			if req.SeqNum > numObservedDonTimes {
+				p.store.requests.Evict(req.WorkflowExecutionID)
+				req.SendResponse(ctx,
+					DonTimeResponse{
+						WorkflowExecutionID: req.WorkflowExecutionID,
+						SeqNum:              req.SeqNum,
+						Timestamp:           0,
+						Err: fmt.Errorf("requested seqNum %d for executionID %s is greater than the number of observed don times %d",
+							req.SeqNum, req.WorkflowExecutionID, numObservedDonTimes),
+					})
+				continue
+			}
+
+			requests[req.WorkflowExecutionID] = int64(req.SeqNum)
+			batchOffset++
+		}
+	}
+
+	observation := &pb.Observation{
+		Timestamp: time.Now().UTC().UnixMilli(),
+		Requests:  requests,
+		Finished:  unscheduledFinishedExecutionIDs,
+	}
+
+	return proto.Marshal(observation)
+}
+
+func (p *Plugin) ValidateObservation(_ context.Context, oc ocr3types.OutcomeContext, _ types.Query, ao types.AttributedObservation) error {
+	return nil
+}
+
+func (p *Plugin) ObservationQuorum(_ context.Context, _ ocr3types.OutcomeContext, _ types.Query, aos []types.AttributedObservation) (quorumReached bool, err error) {
+	return quorumhelper.ObservationCountReachesObservationQuorum(quorumhelper.QuorumTwoFPlusOne, p.config.N, p.config.F, aos), nil
+}
+
+func (p *Plugin) Outcome(_ context.Context, outctx ocr3types.OutcomeContext, _ types.Query, aos []types.AttributedObservation) (ocr3types.Outcome, error) {
+	observationCounts := map[string]int64{} // counts how many nodes reported where a new DON timestamp might be needed
+	finishedNodes := map[string]int64{}     // counts number of nodes finished with the workflow for executionID
+	var times []int64
+
+	prevOutcome := &pb.Outcome{}
+	if err := proto.Unmarshal(outctx.PreviousOutcome, prevOutcome); err != nil {
+		return nil, err
+	}
+
+	for _, ao := range aos {
+		observation := &pb.Observation{}
+		if err := proto.Unmarshal(ao.Observation, observation); err != nil {
+			return nil, err
```

Comment:
This is even more problematic - if one malicious notes sends us crap, we fail the whole plugin. We have to ignore the error here and continue.